src/net.cpp

```diff
@@ -699,37 +723,40 @@ const uint256& V1TransportDeserializer::GetMessageHash() const
     return data_hash;
 }
 
-CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time) {
+Optional<CNetMessage> V1TransportDeserializer::GetMessage(int64_t time)
+{
     // decompose a single CNetMessage from the TransportDeserializer
     CNetMessage msg(std::move(vRecv));
+    Optional<CNetMessage> result{};
```

Comment:
Alternatively, rather than having both a `CNetMessage` and `Optional<CNetMessage>`:

```diff
diff --git a/src/net.cpp b/src/net.cpp
index c23cf54c75..7cc5068a75 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -726,14 +726,13 @@ const uint256& V1TransportDeserializer::GetMessageHash() const
 Optional<CNetMessage> V1TransportDeserializer::GetMessage(int64_t time)
 {
     // decompose a single CNetMessage from the TransportDeserializer
-    CNetMessage msg(std::move(vRecv));
-    Optional<CNetMessage> result{};
+    Optional<CNetMessage> msg(std::move(vRecv));
 
     // store command string, time, and sizes
-    msg.m_command = hdr.GetCommand();
-    msg.m_time = time;
-    msg.m_message_size = hdr.nMessageSize;
-    msg.m_raw_message_size = hdr.nMessageSize + CMessageHeader::HEADER_SIZE;
+    msg->m_command = hdr.GetCommand();
+    msg->m_time = time;
+    msg->m_message_size = hdr.nMessageSize;
+    msg->m_raw_message_size = hdr.nMessageSize + CMessageHeader::HEADER_SIZE;
 
     uint256 hash = GetMessageHash();
 
@@ -743,20 +742,20 @@ Optional<CNetMessage> V1TransportDeserializer::GetMessage(int64_t time)
     // Check checksum and header command string
     if (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) != 0) {
         LogPrint(BCLog::NET, "CHECKSUM ERROR (%s, %u bytes), expected %s was %s, peer=%d\n",
-            SanitizeString(msg.m_command), msg.m_message_size,
+            SanitizeString(msg->m_command), msg->m_message_size,
             HexStr(hash.begin(), hash.begin() + CMessageHeader::CHECKSUM_SIZE),
             HexStr(hdr.pchChecksum, hdr.pchChecksum + CMessageHeader::CHECKSUM_SIZE),
             node_id);
+        msg = nullopt;
     } else if (!hdr.IsCommandValid()) {
         LogPrint(BCLog::NET, "HEADER ERROR - COMMAND (%s, %u bytes), peer=%d\n",
-            hdr.GetCommand(), msg.m_message_size, node_id);
-    } else {
-        result = std::move(msg);
+            hdr.GetCommand(), msg->m_message_size, node_id);
+        msg = nullopt;
     }
 
     // Always reset the network deserializer (prepare for the next message)
     Reset();
-    return result;
+    return msg;
 }
```

Reply:
I'm happy with this pattern if everyone else is (@jonatack).  It's the "reverse" of before.  Now we're assuming success and have to set failure.  I think this is clearer.

Reply:
LGTM

Reply:
LGTM